File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean find(TreeNode root,int sum,HashSet<Integer>H){
        
        if(root==null){
            
            return false;
        }
        
      
        if(H.contains((sum-root.val))){
            
         //   System.out.println(sum+" "+root.val);
            
            return true;
        }
        
        boolean answer=false;
        
        if(root.right!=null&&root.right.val>sum){
            
            H.add(root.val);
            
            answer|=find(root.left,sum,H);
            
            answer|=find(root.right,sum,H);
        }
        else{
            
            H.add(root.val);
            
            answer|=find(root.left,sum,H);
            
            answer|=find(root.right,sum,H);
        }
        
        return answer;
    }
    public boolean findTarget(TreeNode root, int k) {
        
        boolean find=find(root,k,new HashSet());
        
        return find;
    }
}

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
  
 def __binarySearch(self, root, k):
    if root is None:
      return

    if root.val == k:
      return root
    elif root.val > k:
      return self.__binarySearch(root.left, k)
    else:
      return self.__binarySearch(root.right, k)

 def __findTargetBST(self, node, root, k):
         
    if node is None:
      return False

    complement = k - node.val
    comp_node = self.__binarySearch(root, complement)
    if comp_node is not None and node is not comp_node:
      return True

    return self.__findTargetBST(node.left, root, k) or self.__findTargetBST(node.right, root, k)

 def findTarget(self, root, k):
    return self.__findTargetBST(root, root, k)
        